components/style/properties_and_values.rs

```diff
@@ -0,0 +1,1167 @@
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        dest.write_str(&*self.0)
+    }
+
+    #[cfg(feature = "gecko")]
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        dest.write_str(&self.0.to_string())
+    }
+}
+
+impl Deref for Ident {
+    type Target = Atom;
+
+    fn deref(&self) -> &Atom {
+        &self.0
+    }
+}
+
+/// A computed <url> value.
+///
+/// FIXME: While as_str() resolves URLs on the Servo side, it currently does not
+/// resolve them on the Gecko side (see the comment on Gecko's
+/// specified::url::SpecifiedUrl implementation). That means we don't actually
+/// resolve URLs when running in Stylo yet.
+#[derive(Clone, Debug, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct ComputedUrl(pub String);
+
+impl ComputedUrl {
+    #[cfg(feature = "servo")]
+    fn from_specified(url: &specified::url::SpecifiedUrl) -> Result<ComputedUrl, ()> {
+        url.url().map(|x| ComputedUrl(x.as_str().to_owned())).ok_or(())
+    }
+
+    #[cfg(feature = "gecko")]
+    fn from_specified(url: &specified::url::SpecifiedUrl) -> Result<ComputedUrl, ()> {
+        // Doesn't work.
+        // url.extra_data.join(url.as_str()).map(|x| ComputedUrl(x.as_str().to_owned())).ok_or(())
+        Ok(ComputedUrl(url.as_str().to_owned()))
+        // Note: the Gecko binding doesn't currently output resolved URLs.
+    }
+}
+
+impl ToCss for ComputedUrl {
+    fn to_css<W>(&self, dest: &mut W) -> fmt::Result where W: fmt::Write {
+        dest.write_str("url(")?;
+        self.0.to_css(dest)?;
+        dest.write_str(")")
+    }
+}
+
+/// A basic custom property syntax string for a custom property that, used to
+/// build up disjunctions and list terms.
+#[derive(Debug, PartialEq)]
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub enum Type {
+    /// Syntax to allow any valid <length> value.
+    Length,
+    /// Syntax to allow any valid <number> value.
+    Number,
+    /// Syntax to allow any valid <percentage> value.
+    Percentage,
+    /// Syntax to allow any valid <length> or <percentage> value, or any valid
+    /// <calc()> expression combining <length> and <percentage> components.
+    LengthPercentage,
+    /// Syntax to allow any valid <color> value.
+    Color,
+    /// Syntax to allow any valid <image> value.
+    Image,
+    /// Syntax to allow any valid <url> value.
+    Url,
+    /// Syntax to allow any valid <integer> value.
+    Integer,
+    /// Syntax to allow any valid <angle> value.
+    Angle,
+    /// Syntax to allow any valid <time> value.
+    Time,
+    // FIXME: We should support <resolution> values as well.
+    /// Syntax to allow any valid <transform-list> value.
+    TransformList,
+    /// Syntax to allow any valid <custom-ident> value.
+    CustomIdent,
+    /// Syntax to allow a specific identifier (matching the <custom-ident>
+    /// production, compared codepoint-wise).
+    SpecificIdent(Ident),
```

Comment:
Oh, I guess this would match if I used a syntax like `foo bar`, then I'd have two `SpecificIdent`s, right?